### ISSUE TITLE
Segmentation fault (signal 11)

### ISSUE THREAD

**Issue by sv-bio** [spyking-circus/spyking-circus#332]
Dear Pierre,

Recently I noticed rather strange behavior of the SC:
During processing of some of my recordings (not all, but they are all about 20 min-long, from the same animal) often, but not always I get the following error:

File          : /path/to/data/file/data.dat
Steps         : filtering, whitening, clustering, fitting, merging
Number of CPU : 4/4
Parallel HDF5 : False
Shared memory : True
Hostfile      : /path/to/spyking-circus/circus.hosts

Informations  -------------------------
| Number of recorded channels : 32
| Number of analyzed channels : 32
| File format                 : RAW_BINARY
| Data type                   : float32
| Sampling rate               : 30000 Hz
| Duration of the recording   : 20 min 1 s 105 ms
| Width of the templates      : 2 ms
| Spatial radius considered   : 60 um
| Threshold crossing          : negative
------------------------------------------------------------------
-------------------------  Informations  -------------------------
| Filtering has already been done
| Median over all channels has already been removed
------------------------------------------------------------------
Analyzing data to get whitening matrices and thresholds...
Found 30s to compute the whitening matrix...
Because of whitening, need to recompute the thresholds...
Searching spikes to construct the PCA basis...
100%|████████████████████████████████████|[01:21<00:00, 20.33s/it]
Found 25600 waveforms over 25600 requested
-------------------------  Informations  -------------------------
| A basis with 5 dimensions has been built
------------------------------------------------------------------
Searching isolated random spikes to sample amplitudes...
100%|████████████████████████████████████|[01:18<00:00, 19.59s/it]
Found 256000 spikes over 256000 requested
Estimating amplitudes distributions...
Smart Search of good isolated spikes for the clustering (1/3)...
100%|███████████████████████████████████|[07:20<00:00, 110.10s/it]
Found 171863 isolated spikes over 256000 requested (1595988 rejected)
Computing density estimations...
Searching random spikes to refine the clustering (2/3)...
100%|████████████████████████████████████|[01:10<00:00, 17.53s/it]
Found 256000 spikes over 256000 requested
Refining density estimations...
Searching random spikes to refine the clustering (3/3)...
100%|████████████████████████████████████|[01:22<00:00, 20.65s/it]
Found 256000 spikes over 256000 requested
Refining density estimations...
Running density-based clustering...
100%|████████████████████████████████████|[01:22<00:00, 10.34s/it]
-------------------------  Informations  -------------------------
| Number of clusters found : 303
| Number of local merges   : 0 (method nd-bhatta, param 2)
------------------------------------------------------------------
Estimating the templates with the median-raw procedure ...
100%|████████████████████████████████████|[00:52<00:00,  6.62s/it]
Removing mixtures of templates...
100%|████████████████████████████████████|[00:00<00:00, 19.17it/s]
  0%|                                            |[00:00<?, ?it/s]
===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 2329 RUNNING AT pharma
=   EXIT CODE: 11
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Segmentation fault (signal 11)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
----------------------------  Error  -----------------------------
| Step "clustering" failed for reason Command '['mpiexec', '-np', '4', 'spyking-circus-subtask', 'clustering', '/path/to/data/file/data.dat', '4', '0', 'False', 'False']' returned non-zero exit status 11.!

Sometimes the same file may be processed without any issues. If I process recording, that consists of several chunks of data (2 in my case) I could only analyze 1 file at a time because if stream_mode = multi-files, high-amplitude peaks on some channels became unrecognised:

![Screenshot_2020-12-14_02-51-49](https://user-images.githubusercontent.com/58944642/102029511-5dfb6c80-3db7-11eb-8b0c-93e859283623.png)

and these channels appear empty in Merging GUI:

![Screenshot_2020-12-14_02-43-43](https://user-images.githubusercontent.com/58944642/102029448-33a9af00-3db7-11eb-9d87-6e62a709f71a.png)

I observed similar behaviour when I reduce the number of CPU cores to 1 during processing one file at a time. I thought this could be related to memory allocation, but at a time the error evokes only 20%  RAM is used.

I tried shared_memory = False (as described in issue #76), but with no success. To my surprise, sanity plots had templates on those channels without peaks.

Parameter file:
[parameters.txt](https://github.com/spyking-circus/spyking-circus/files/5685900/parameters.txt)

The system is:
Debian 10.4
SC 1.0.6

Thank you for help in advance.

Sincerely,
Stan

**Comment by yger** [comment]
No problem, thanks for reporting. But again, sorry to bother, but can you upload a recording/param/probe? Always better to debug. I've no idea why some channels are missing.... Regarding the bug with the mixtures, I guess if this is only this method which is faulty, it can temporarily be avoided by setting remove_mixture = False in [clustering]

**Comment by sv-bio** [comment]
I attached log file for the post above. remove_mixture = False helped with at least one recording.
I'll close the issue after a few more  trials.
Thanks.

**Comment by yger** [comment]
Helping is not solving the problems... If you can provide me the data I could try to understand :
- why the remove_mixture step is crashing
- why some spikes are not detected, while they clearly should be

**Comment by sv-bio** [comment]
I sent data file+results of the analysis+log via e-mail.

**Comment by sv-bio** [comment]
Shall I close this issue and open separate one about missing channels/templates?
I also have some more recordings with missing channels/templates. I can share these data with you if necessary.

**Comment by yger** [comment]
With the data you gave me, I can not observe the bahvior. I should with the default param, ma I right? I.e. crash during mixtures and/or missing channels? Leave it open, I need to understand if this is also windows related

**Comment by sv-bio** [comment]
Do you see missing channels/templates at Merging GUI and result window?

**Comment by yger** [comment]
I ran the algorithm, and see normal results. But I'll look at yours then

**Comment by yger** [comment]
Actually, just to be sure we are in sync. When you say "missing channels", it means that when you make a selection of templates, sometimes, they do not cover the whole spaces of the channels. The channels not involved in your template selection are not displayed. So based on your selection, you can see only a subset of the channels. Be also sure that the box "Show only peaks is not ticked". Otherwise, the GUi is only displaying, for every template, the channels where it is peaking (and not all where it is defined)

**Comment by sv-bio** [comment]
Yes, by saying "missing channels" I mean exactly this (it looks like fig. 2 of my first post in this thread). When I uncheck ""Show only peaks" missing channels appear as flat lines and the "results" view looks same as figure 1 above. Do you think it is OS-specific issue? In this case, versions of what libraries/packages shall I check?

**Comment by yger** [comment]
No, but then everything is normal. Missing channels appears as flat lines means that there is no info on these nearby channels, so the code sparsified everything setting O

**Comment by sv-bio** [comment]
Surely. I re-run the whole analysis every time I modify any parameter in*.params file.

**Comment by yger** [comment]
No, the templates are defined on only 8 channels at max (because they are restricted to shanks), and then, within this 8 channels, some are set to 0 when not enough information is present, in order to sparsify the templates. This is what you can see when you look at the results in phy: the templates shown in phy are exactly the ones used in SC, so you can see that they have variable sizes for the spatial extend, but always restricted within a shank

**Comment by sv-bio** [comment]
I see. Meanwhile to check whether this Error 11 and absence of fitted spikes on some of the channels is caused by some particular channel I  re-run analysis with the same settings, but left only selected channels, marking others as bad. I found that Error 11 never occurs and all the channels have spikes detected when I analyze up to 24 channels at once. When there are more than 24 channels, Error 11 appears sporadiacally with no clear pattern i.e. I may get it during one run but succeed with the same settings and data on the other, so probably it is specific to my work station. As of unfitted peaks, I found that the more channels and templates I have and the longer the recording the more channels have unfitted peaks (a on figure 2 above). For now I will analyze my data shank-wise, but will still try to figure out why peaks on some channels remain unfitted. Please, let me know if you have an idea to try or solution for it.
Sincerely,
Stan

**Comment by sv-bio** [comment]
I think, this Segmentation fault (signal 11) and channels without detected spikes (like in fig. 1 and 2 above) may have something to do with RAM for the following reason. Each of my data sets consists of 8*4 channels, 30 kHz sampling rate, and recorded in 3 files, 20 min each totalling to 60 min. I set templates_normalization = False because of issue #329.  When I try to analyse such complete set (8*4 ch * 30 kHz * 60 min) Segmentation fault (signal 11)  appears. When, instead, I leave only 8*3 ch *  30 kHz * 60 min, analysis completes, but some channels have no detected spikes. If I run 8*4 ch *  30 kHz * 20 min or  8*1 ch * 30 kHz * 60 min, the analysis completes smoothly and spikes are found on all the channels, as expected. Because of templates_normalization = False I have 100+ templates which, in conjunction with my data size and limited computational resources (16 Gb RAM + 5 Gb swap), could cause such behavior. What do you think? As a workaround I run the analysis on each shank (8 channels) separately.

**Comment by yger** [comment]
Not sure this is related to RAM. Back from holidays, I'll try to solve your problem, but on my machine (16Go of RAM), on a first glimpse launching the code on the data you gave me works normally. However, I'm getting a bit lost... Can you remind me what are the files I should look for?

**Comment by sv-bio** [comment]
First part of the recording: Rozov_1__uid0928-14-50-02.dat
second part: Rozov_2__uid0928-15-10-03.dat
